docs/readmes/proposals/p006_mandatory_integration_tests_for_each_PR.md

```diff
@@ -0,0 +1,62 @@
+ - magma_integration_tests pipeline takes ~1 hour to execute and Jenkins can process only 1 run/hour (does not support concurrent builds)
+ - average magma_integration_tests queue length was 30-40 jobs (before opt-in strategy was implemented)
+ - CWAG-integration-test and LTE-integration-test jobs take 1 hour each (before optimisation.) Jenkins can process approximtely 13 executions per hour. 
+ Recently we switched to "Opt-In strategy" for PRs affecting the gateway code base. These PRs must be marked with one of the following labels: "component: agw" OR "component: cwf" OR "component: feg". Implementation of this strategy decreased load on jenkins (went from 30+ to 2-10 per day). At the same time, this means that PRs which were not marked with the correct label are not tested and may be introducing new bugs.
+ 
+ We can draw a conclusion from above and outline the following pain points:
+ - Some PRs are still merged into master without integration tests
+ - Integration tests are slow (actual execution is slow, regardless of VM setup time) so we can't realistically run them for each commit of each PR.
+ - Reviewers are not applying tags to PRs and waiting for them to complete (or skipping gatekeeping tests)
+ - Magma_integration_tests pipeline is the largest bottleneck in our CI process
+ 
+
+## Proposed solutions
+
+### Integration tests are slow(which is more or less expected) so we can't run them for each commit of each PR.
+We should execute integration tests for each and every PR which is going to be merged in to master. But there's no sense to do it for each and every commit.
+Solution 
+Stage 1:
+ - Ensure that for each commit of each PR all unit test jobs are executed by CircleCI
+ - Ensure that each PR was reviewed and accepted by component maintainer
+ - Ensure that each PR which succesfully passed unit testing and maintainer review gets "ready-to-merge" label assigned.
+ - Ensure that each PR marked with label "ready-to-merge" was tested by all integration tests.
+ Stage 2:
+ - Ensure that each PR marked with label "ready-to-merge", which passed integration tests is automatically merged by CI.
+ Stage 3:
+ - Ensure that no side effects appear when PRs are automatically merged by CI (so speculative CI implementation in fact).
+
+### Some PRs are still merged into master without integration tests
+Make Jenkins as a mandatory gatekeeper for merging each PR. Proposed workflow:
+ - Reviewer marks appropriate labels to run necessary relevant tests (such as CWAG / NMS / etc.) before indicating "ready-to-merge" (tag or some keyword in a comment)
+ - Jenkins triggers all tests (or if we decide, just a subset)
```

Comment:
We should start with all integration tests and then try and understand why we will reduce the surface area if needed.

Reply:
make sense - removed this comment